core/trino-main/src/test/java/io/trino/TestSystemSessionProperties.java

```diff
@@ -0,0 +1,95 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino;
+
+import io.airlift.units.Duration;
+import io.trino.execution.DynamicFilterConfig;
+import io.trino.execution.QueryManagerConfig;
+import io.trino.execution.TaskManagerConfig;
+import io.trino.execution.scheduler.NodeSchedulerConfig;
+import io.trino.memory.MemoryManagerConfig;
+import io.trino.memory.NodeMemoryConfig;
+import io.trino.spi.session.PropertyMetadata;
+import io.trino.sql.planner.OptimizerConfig;
+import org.testng.annotations.Test;
+
+import static io.trino.SystemSessionProperties.QUERY_MAX_RUN_TIME;
+import static java.util.concurrent.TimeUnit.HOURS;
+import static org.assertj.core.api.Assertions.assertThatNoException;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class TestSystemSessionProperties
+{
+    @Test
+    public void testQueryMaxRunTimeWithoutHardLimit()
+    {
+        SystemSessionProperties properties = new SystemSessionProperties(
+                new QueryManagerConfig()
+                        .setQueryMaxRunTime(new Duration(2, HOURS)),
+                new TaskManagerConfig(),
+                new MemoryManagerConfig(),
+                new FeaturesConfig(),
+                new OptimizerConfig(),
+                new NodeMemoryConfig(),
+                new DynamicFilterConfig(),
+                new NodeSchedulerConfig());
+        PropertyMetadata<Duration> property = getProperty(properties, QUERY_MAX_RUN_TIME);
+        assertThatNoException().isThrownBy(() -> property.decode("1h"));
+        assertThatNoException().isThrownBy(() -> property.decode("3h"));
+    }
+
+    @Test
+    public void testQueryMaxRunTimeWithHardLimit()
+    {
+        SystemSessionProperties properties = new SystemSessionProperties(
+                new QueryManagerConfig()
+                        .setQueryMaxRunTime(new Duration(2, HOURS))
+                        .setQueryMaxRunTimeHardLimit(new Duration(4, HOURS)),
+                new TaskManagerConfig(),
+                new MemoryManagerConfig(),
+                new FeaturesConfig(),
+                new OptimizerConfig(),
+                new NodeMemoryConfig(),
+                new DynamicFilterConfig(),
+                new NodeSchedulerConfig());
+        PropertyMetadata<Duration> property = getProperty(properties, QUERY_MAX_RUN_TIME);
+        assertThatNoException().isThrownBy(() -> property.decode("1h"));
+        assertThatNoException().isThrownBy(() -> property.decode("3h"));
+        assertThatThrownBy(() -> property.decode("5h"))
+                .hasMessage("query_max_run_time must not exceed 'query.max-run-time.hard-limit' 4.00h: 5.00h");
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T> PropertyMetadata<T> getProperty(SystemSessionProperties properties, String propertyName)
```

Comment:
You can get rid of warning suppression by declaring the method as returning `PropertyMetadata<?>`.

Reply:
```suggestion
    private static PropertyMetadata<?> getProperty(SystemSessionProperties properties, String propertyName)
```

Reply:
Done